### PR TITLE
Add support for reading B3 single header

### PR DESCRIPTION
**Pull request by ykitamura-mdsol**
This PR is just to support the "reading" part of B3 single header (#126).
The writing part is not included.

@adriancole sorry for taking a long time but I like the single header!

@jcarres-mdsol @jfeltesse-mdsol

### CODE REVIEW

File: lib/zipkin-tracer/zipkin_b3_single_header_format.rb [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+    def to_a
+      [trace_id, span_id, parent_span_id, sampled, flags]
+    end
+
+    def self.parse_from_header(b3_single_header)
+      return new(parse_sampled_flags(b3_single_header)) if b3_single_header.size == 1
+
+      trace_id, span_id, flag, parent_span_id = b3_single_header.split('-')
+      new(trace_id: trace_id, span_id: span_id, parent_span_id: parent_span_id, **parse_sampled_flags(flag))
```
Comment (jfeltesse-mdsol) [inline lib/zipkin-tracer/zipkin_b3_single_header_format.rb]: I think it's a bit wasteful to have this method create a new array (by calling split), then a new object and then in the caller call `to_a` which is going to create yet another array.

How about having a method getting the header as an input and return an array directly, without instantiating?

File: lib/zipkin-tracer/zipkin_b3_single_header_format.rb [hunk 2]
```diff
@@ -0,0 +1,38 @@
+module ZipkinTracer
```
Comment (jfeltesse-mdsol) [inline lib/zipkin-tracer/zipkin_b3_single_header_format.rb]: how about adding the magic comment about frozen strings?

File: lib/zipkin-tracer/rack/zipkin_env.rb [hunk 3]
```diff
@@ -10,33 +10,44 @@ def initialize(env, config)
...
+
     def called_with_zipkin_headers?
       @called_with_zipkin_headers ||= B3_REQUIRED_HEADERS.all? { |key| @env.key?(key) }
     end
 
     private
 
-    B3_REQUIRED_HEADERS = %w(HTTP_X_B3_TRACEID HTTP_X_B3_SPANID).freeze
-    B3_OPT_HEADERS = %w(HTTP_X_B3_PARENTSPANID HTTP_X_B3_SAMPLED HTTP_X_B3_FLAGS).freeze
+    B3_SINGLE_HEADER = 'HTTP_B3'.freeze
```
Comment (adriancole) [inline lib/zipkin-tracer/rack/zipkin_env.rb]: easier to quote, ain't it :)

File: lib/zipkin-tracer/zipkin_b3_single_header_format.rb [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+  # This format corresponds to the propagation key "b3" (or "B3").
+  # b3: {x-b3-traceid}-{x-b3-spanid}-{if x-b3-flags 'd' else x-b3-sampled}-{x-b3-parentspanid}
+  # For details, see: https://github.com/apache/incubator-zipkin-b3-propagation
```
Comment (adriancole) [inline lib/zipkin-tracer/zipkin_b3_single_header_format.rb]:
```suggestion
  # For details, see: https://github.com/openzipkin/b3-propagation
```

File: lib/zipkin-tracer/zipkin_b3_single_header_format.rb (hunk 1 above)
Comment (adriancole) [reply lib/zipkin-tracer/zipkin_b3_single_header_format.rb]: though I would say this is the least code approach I've ever seen :P

File: spec/lib/zipkin_b3_single_header_format_spec.rb [hunk 5]
```diff
@@ -0,0 +1,113 @@
+require 'spec_helper'
+
+describe ZipkinTracer::B3SingleHeaderFormat do
+  let(:b3_single_header_format) { described_class.parse_from_header(b3_single_header) }
+
+  context 'child span' do
```
Comment (adriancole) [inline spec/lib/zipkin_b3_single_header_format_spec.rb]: add some malformed tests? Ex in brave we use

"not-a-tumor"
"b970dafd-0d95-40aa-95d8-1d8725aebe40"

File: lib/zipkin-tracer/zipkin_b3_single_header_format.rb (hunk 2 above)
Comment (ykitamura-mdsol) [reply lib/zipkin-tracer/zipkin_b3_single_header_format.rb]: Added in f8d6a89

File: lib/zipkin-tracer/zipkin_b3_single_header_format.rb (hunk 1 above)
Comment (ykitamura-mdsol) [reply lib/zipkin-tracer/zipkin_b3_single_header_format.rb]: Updated in 2fef91b

File: spec/lib/zipkin_b3_single_header_format_spec.rb (hunk 5 above)
Comment (ykitamura-mdsol) [reply spec/lib/zipkin_b3_single_header_format_spec.rb]: @adriancole In ruby client, we are doing malformed check in another class: 

https://github.com/openzipkin/zipkin-ruby/blob/a1b1523a5bfcde3b6d2a8213a76e76b7a0e8a088/lib/zipkin-tracer/trace.rb#L134-L150

I think we don't need to have malformed tests here but what do you think?

**Comment by adriancole** [comment]
sounds fine

**Pull request merged**